sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs

```diff
@@ -0,0 +1,230 @@
+                        }
+                    }
+                },
+                SyntaxKind.InvocationExpression);
+        }
+
+        private static bool IsParentMethodContainsEndInvoke(SyntaxNode node, SemanticModel semantic)
+        {
+            var parentContext = node.AncestorsAndSelf()
+                .FirstOrDefault(ancestor => ParentDeclarationKinds.Contains(ancestor.Kind()));
+            return GetEndInvokeList(parentContext, semantic).Count > 0;
+        }
+
+        private ExpressionSyntax GetCallbackArg(InvocationExpressionSyntax invocationExpression)
+        {
+            var callbackArgPos = invocationExpression.ArgumentList.Arguments.Count - 2;
+            var callbackArg = GetArgumentExpressionByNameOrPosition(invocationExpression, "callback", callbackArgPos)
+                ?.RemoveParentheses();
+            return callbackArg;
+        }
+
+        private static bool IsCallbackMayContainEndInvoke(SyntaxNode callbackArg, SemanticModel semantic)
+        {
+            var calledMethod = callbackArg.RemoveParentheses();
+            calledMethod = ReplaceIdentifierByItsInitializer(calledMethod, semantic);
+            calledMethod = ReplaceNewDelegateByMethodDeclaration(calledMethod, semantic);
+            if (calledMethod != null && ParentDeclarationKinds.Contains(calledMethod.Kind()))
+            {
+                return GetEndInvokeList(calledMethod, semantic).Count > 0;
+            }
+
+            return true;
+        }
+
+        private static SyntaxNode ReplaceNewDelegateByMethodDeclaration(SyntaxNode node, SemanticModel semantic)
+        {
+            if (node != null && node.IsKind(SyntaxKind.ObjectCreationExpression))
+            {
+                var arguments = ((ObjectCreationExpressionSyntax)node).ArgumentList.Arguments;
+                if (arguments.Count == 1 &&
+                    semantic.GetSymbolInfo(arguments[0].Expression).Symbol is IMethodSymbol symbol)
+                {
+                    return symbol?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax();
+                }
+            }
+
+            return node;
+        }
+
+        private static SyntaxNode ReplaceIdentifierByItsInitializer(SyntaxNode node, SemanticModel semantic)
+        {
+            if (node != null && node.IsKind(SyntaxKind.IdentifierName))
+            {
+                var declaringReference =
+                    semantic.GetSymbolInfo(node).Symbol?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax();
+                if (declaringReference is VariableDeclaratorSyntax variableDeclarator
+                    && variableDeclarator.Initializer is EqualsValueClauseSyntax equalsValueClause)
+                {
+                    return equalsValueClause.Value.RemoveParentheses();
+                }
+            }
+
+            return node;
+        }
+
+        private static bool IsNullLiteral(ExpressionSyntax callbackArg) => callbackArg.IsKind(SyntaxKind.NullLiteralExpression);
```

Comment:
I think this could be called on a `SyntaxNode`, and you could make it an extension method (see the `Helpers` namespace to add it to SyntaxNodeHelpers

Reply:
I didn't find `SyntaxNodeHelpers` but `SonarAnalyzer.CSharp\Helpers\CSharpSyntaxHelper`. I hope it's what you wanted to say. I created an extension method.